Document/MessageManager.php

```diff
@@ -0,0 +1,232 @@
+use NotificationEngine\Sonata\NotificationBundle\Model\MessageInterface;
+use NotificationEngine\Sonata\NotificationBundle\Model\MessageManagerInterface;
+use Sonata\CoreBundle\Model\BaseDocumentManager;
+use Sonata\DatagridBundle\Pager\Doctrine\Pager;
+use Sonata\DatagridBundle\ProxyQuery\Doctrine\ProxyQuery;
+
+/**
+ * @author Salma Khemiri <chakroun.salma@gmail.com>
+ */
+class MessageManager extends BaseDocumentManager implements MessageManagerInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function save($message, $andFlush = true)
+    {
+        //Hack for ConsumerHandlerCommand->optimize()
+        if ($message->getId() && !$this->dm->getUnitOfWork()->isInIdentityMap($message)) {
+            $this->getDocumentManager()->getUnitOfWork()->merge($message);
+        }
+
+        parent::save($message, $andFlush);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function findByTypes(array $types, $state, $batchSize)
+    {
+        $query = $this->prepareStateQuery($state, $types, $batchSize);
+
+        $result = $query->getQuery()->execute();
+
+        if ($result instanceof Cursor) {
+            $result = $result->toArray();
+        }
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function findByAttempts(array $types, $state, $batchSize, $maxAttempts = null, $attemptDelay = 10)
+    {
+        $query = $this->prepareStateQuery($state, $types, $batchSize);
+
+        if ($maxAttempts) {
+            $now = new \DateTime();
+            $delayDate = $now->add(\DateInterval::createFromDateString(($attemptDelay * -1).' second'));
+
+            $query
+                ->field('restartCount')->lt($maxAttempts)
+                ->field('updatedAt')->lt($delayDate);
+        }
+
+        $result = $query->getQuery()->execute();
+
+        if ($result instanceof Cursor) {
+            $result = $result->toArray();
+        }
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function countStates()
+    {
+        $result = $this->getRepository()->createQueryBuilder('message')
+                                        ->group(array('state' => 1), array('count' => 0))
+                                        ->reduce('function (curr, result) { result.count++; }')
+                                        ->getQuery()
+                                        ->execute();
+
+        $states = array(
+            MessageInterface::STATE_DONE => 0,
+            MessageInterface::STATE_ERROR => 0,
+            MessageInterface::STATE_IN_PROGRESS => 0,
+            MessageInterface::STATE_OPEN => 0,
+        );
+
+        if ($result instanceof ArrayIterator) {
+            $result = $result->toArray();
+
+            foreach ($result as $data) {
+                $states[$data['state']] = $data['count'];
+            }
+        }
+
+        return $states;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function cleanup($maxAge)
```

Comment:
I think `maxAge` can't be a datetime, it's a duration, not a date. Building a `\DateInterval` is quite cumbersome so this is fine IMO.